### ISSUE TITLE
Java Warning: An illegal reflective access operation has occurred when recursive comparison with compiled Regex Pattern

### ISSUE THREAD

**Issue by achterberatung** [assertj/assertj-core#2254]
#### Summary

I have a test, where I use Mockito Captor to capture the argument for a function of a mocked class.

This function gets an object with a compiled regex pattern inside.

I manually build an expected object on the same way and assert equality with assert4j.

So, everything works fine and the comparison fails, if I change my expected regex, but Java puts out this warning, when running the test:

/usr/lib64/jvm/java-11-openjdk-11/bin/java ...

WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.assertj.core.util.introspection.FieldUtils (file:/..../repository/org/assertj/assertj-core/3.18.1/assertj-core-3.18.1.jar) to field java.util.regex.Pattern.cursor
WARNING: Please consider reporting this to the maintainers of org.assertj.core.util.introspection.FieldUtils
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release

I am not sure, if it is a problem with org.mongojack.DBQuery (but that's only some value holding object), with Mockito or with assertj, but Java told me, to report it to you :-)  

#### Example

I hope the example gives you a clue, I unfortunately don't have the time, to build an independend and fully working/reproducing example.

```java
import org.mongojack.DBQuery;

...

@Captor
private ArgumentCaptor<DBQuery.Query> queryCaptor;

...

DBQuery.Query expectedFilter = DBQuery.and(
                DBQuery.regex("id", Pattern.compile(".*((?i)aaa123).*", Pattern.CASE_INSENSITIVE)),
                DBQuery.regex("textfeld", Pattern.compile(".*((?i)Haha).*", Pattern.CASE_INSENSITIVE)));
when(mockedClass.someFunction(queryCaptor.capture())).thenReturn("result");

testedClass.doStuff();

assertThat(queryCaptor.getValue()).usingRecursiveComparison().isEqualTo(expectedFilter);

```

**Comment by achterberatung** [comment]
Hi there,

i have found the reason by myself. It is all about the compiled Regexes, which are producing the warning.

As i use org.mongojack.DBQuery, now replaced by org.bson.conversions.Bson (which have the same interface), i can use the function .toBsonDocument() from the BSON objects. These are comparable and my tests are now fully working.

`        assertThat(queryCaptor.getValue().toBsonDocument()).usingRecursiveComparison().isEqualTo(expectedQuery.toBsonDocument());
`

**Comment by scordio** [comment]
Related to #2466 and #2450, fixed with 7406133f21ed66e3908a50c581ad0f0aaf059aa6.

Will be part of 3.23.0.
